Add support for ffmpeg's segment deletion for HLS

DomiStyle opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our documentation.
-->

**Changes**
This PR adds a new setting to enable segment deletion and wrapping in ffmpeg for HLS playback. This prevents the cache drive from filling up for long transcode sessions and allows using a small ramdisk for the transcoding files.

The option is disabled by default and everything reverts to the current behavior when the throttler and segment deletion are not both enabled.

This also fixes the logging for the throttler, which was broken before.

**Issues**
https://github.com/jellyfin/jellyfin/issues/3929

cvium commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1503,8 +1503,17 @@ private async Task<ActionResult> GetDynamicSegment(StreamingRequestDto streaming
                         try
                         {
-                            await _transcodingJobHelper.KillTranscodingJobs(streamingRequest.DeviceId, streamingRequest.PlaySessionId, p => false)
+                            if (_encodingOptions.EnableThrottling && _encodingOptions.EnableSegmentDeletion)
+                            {
+                                // Delete old HLS files when segment deletion is active since ffmpeg doesn't clean them up by itself
+                                await _transcodingJobHelper.KillTranscodingJobs(streamingRequest.DeviceId, streamingRequest.PlaySessionId, p => true)
                                 .ConfigureAwait(false);
+                            }
+                            else
+                            {
+                                await _transcodingJobHelper.KillTranscodingJobs(streamingRequest.DeviceId, streamingRequest.PlaySessionId, p => false)
+                                    .ConfigureAwait(false);
+                            }
```
Simplification. Wouldn't it be safe to delete regardless though?
```suggestion
                            // Delete old HLS files when segment deletion is active since ffmpeg doesn't clean them up by itself
                            var deleteFiles = _encodingOptions.EnableThrottling && _encodingOptions.EnableSegmentDeletion;
                            await _transcodingJobHelper.KillTranscodingJobs(streamingRequest.DeviceId, streamingRequest.PlaySessionId, p => deleteFiles)
                                .ConfigureAwait(false);
```

DomiStyle replied:
> Wouldn't it be safe to delete regardless though?

Not entirely sure, I didn't want to interfere too much with the regular transcoding method since I don't know if the old segments are re-used when scrubbing or not.

cvium commented on Jellyfin.Api/Controllers/DynamicHlsController.cs:
```diff
@@ -1653,11 +1664,38 @@ private string GetCommandLineArguments(string outputPath, StreamState state, boo
+
+                _logger.LogDebug("Using throttling and segment deletion, keeping {0} segments", segmentCount);
+
+                return string.Format(CultureInfo.InvariantCulture, "-hls_list_size {0} -segment_wrap {0} -hls_flags delete_segments", segmentCount.ToString(CultureInfo.InvariantCulture));
```
We typically use `hls_playlist_type vod`, which "Forces hls_list_size to 0; the playlist must not change." (https://ffmpeg.org/ffmpeg-formats.html)

cvium replied:
Without the vod type, you cannot seek ahead of the playlist, so I'm not sure how useful this is.

DomiStyle replied:
> Without the vod type, you cannot seek ahead of the playlist, so I'm not sure how useful this is.

Seeking works perfectly fine. Obviously just within ~3 or so minutes since the throttler is on and outside of it a new transcoder instance is started.

DomiStyle commented:
Looks like the throttler is too aggressive for some files, stalling the stream. Will take a look if it can't be tuned a little bit.

Ideally it would go by segments instead of time but will have to check how exactly it works first.

DomiStyle commented:
Found the issue.

It's a bug with the current transcoder and might cause some of the throttler playback issues. I was able to reproduce this in the Jellyfin 10.8.7 Docker as well.

https://github.com/jellyfin/jellyfin/blob/a214ca259838fa7d1a7052b7bd7ba6a16720406f/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L1283-L1290

These keyframe args sometimes cause ffmpeg to produce segments which are 10 seconds in length instead of the requested length. This obviously causes issues since it completely desyncs the throttler from the actual playback, getting worse the longer you watch. Also breaks seeking the longer a file is.

To reproduce:

* Enable throttler
* Use software transcoding or VAAPI
* Playback a movie starting in the middle
* Transcode it down to 1080p 10 Mbps
* Expected segment length should be 3 seconds
* Actual segment files are 10 seconds in length
* Playback will stall after a few minutes, depending on length of file

I'm not sure yet if this happens under other conditions as well.

I can't test with any of the other hardware encoders except VAAPI, so I can't test how the gopArg reacts yet.

https://github.com/jellyfin/jellyfin/blob/a214ca259838fa7d1a7052b7bd7ba6a16720406f/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L1300-L1303

If I remove the if and use these keyframe args even for vod, everything works:

https://github.com/jellyfin/jellyfin/blob/a214ca259838fa7d1a7052b7bd7ba6a16720406f/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L1276-L1282

I have no idea why ffmpeg is doing that or what the second if branch is even doing, but since @nyanmisaka commited this 2 years ago, maybe he's got an idea?

https://github.com/jellyfin/jellyfin/blame/a214ca259838fa7d1a7052b7bd7ba6a16720406f/MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs#L1276-L1290

Just for reference, this works:
```
-force_key_frames:0 "expr:gte(t,n_forced*3)"
```
```
-force_key_frames:0 "expr:gte(t,0+n_forced*3)"
```
This does not:
```
-force_key_frames:0 "expr:gte(t,363+n_forced*3)"
```

nyanmisaka commented:
> I have no idea why ffmpeg is doing that or what the second if branch is even doing, but since @nyanmisaka commited this 2 years ago, maybe he's got an idea?

These are code inherited from emby 3.5.
Two years ago the `gopArg ` doesn't work on VAAPI so we use `keyFrameArg` to force key frames.
Can you try enable gopArg and disable keyFrameArg? 

Also, what is your iGPU model?

DomiStyle commented:
> Can you try enable gopArg and disable keyFrameArg?

Both `gopArg` and the first variant of `keyFrameArg` works with software and VAAPI encoding. But `gopArg` not as precise as `keyFrameArg`, sometimes producing segments that are 2 or 4 seconds long.
Should be precise enough to not cause issues.

> Also, what is your iGPU model?

No iGPU, I use an AMD 6900 XT for VAAPI.

I just pushed a simpler solution that works for any variation I have tested so far.

nyanmisaka commented:
`-force_key_frames:0 "expr:gte(t,n_forced*3)"`
This change make sense to libx264, libx265 and VAAPI encoders.

But came two new problems based on my test.

First is that deleting segments when killing the transcoder increases the delay of seeking by about 1 to 2 seconds, which is noticeable for click and play.

Second is that the default `SegmentKeepSeconds` is a bit too aggressive (200-180=20s), which causes the segment to be deleted before it is downloaded by the client, resulting in the unnecessary re-encoding of the deleted segment. This problem is more serious in fMP4 HLS, especially when using the very fast hardware encoder (600fps on 1080p transcoding). I changed it to 360 and the issue went away.

DomiStyle commented:
> First is that deleting segments when killing the transcoder increases the delay of seeking by about 1 to 2 seconds, which is noticeable for click and play.

I'm not noticing such a huge difference on my SSD but it definitely is there and probably seriously noticeable on a slow drive.

Not deleting them is a option since usually nobody seeks more than a bunch of times per file.
An alternative would be deleting them in the background? `KillTranscodingJob` could gather a list of files before the new transcoder starts and return early while deleting them in the background.

> I changed it to 360 and the issue went away.

I changed the default to 360, better to have a safe default.

DomiStyle commented:
> Actually I am transcoding to a RAM disk and using localhost to access jellyfin, so the extra latency is noticeable to me. It will be more obvious if hard drive is used as the transcode temp folder.

I reverted to the old behavior for now, that will keep the segment files until the session ends.
Skipping repeatedly will still fill the disk/RAM but it's probably not that big of a deal.

nussdk reviewed (approved):
when/who can approve these changes to merge into master?

RTUnit commented:
There is issue with build for OpenAPI (Generate OpenAPI.json):

```
Error: /home/runner/work/jellyfin/jellyfin/Jellyfin.Api/Helpers/StreamingHelpers.cs(3[54](https://github.com/jellyfin/jellyfin/actions/runs/3472179470/jobs/5802658934#step:4:55),24): error CA1305: The behavior of '{0}' could vary based on the current user's locale settings. Provide a value for the 'IFormatProvider' argument. [/home/runner/work/jellyfin/jellyfin/Jellyfin.Api/Jellyfin.Api.csproj]
Error: /home/runner/work/jellyfin/jellyfin/Jellyfin.Api/Controllers/DisplayPreferencesController.cs(181,29): error CA1305: The behavior of '{0}' could vary based on the current user's locale settings. Provide a value for the 'IFormatProvider' argument. [/home/runner/work/jellyfin/jellyfin/Jellyfin.Api/Jellyfin.Api.csproj]
Error: Process completed with exit code 1.
```
There is possible solution for the problem described here: https://aur.archlinux.org/packages/jellyfin?O=10&PP=10
I have no idea who can try to use/implement it for this GitHub build :/

DomiStyle commented:
@nyanmisaka @cvium Is this feature still wanted? Just asking before I spend any time resolving merge conflicts.

nyanmisaka commented:
@DomiStyle Sorry for the late reply. This looks good to me. Please resolve the merge conflicts.

Shadowghost commented:
Build and tests seem to fail

DomiStyle commented:
Looking into it...

DomiStyle commented:
Build works now, testing playback for another half hour and fixing the web branch as well.

DomiStyle commented:
Ok, finished testing. I picked an even safer default for the segments to keep, it's now 4x the default throttle seconds.
That should prevent throttler runaways even with blazing fast encoders and ramdisk users can tweak the value further to save more RAM.

Web branch is also updated so looks good to me!

Waester commented:
Any chance the throttler fix makes it to the 10.8.x branch?

nielsvanvelzen commented:
10.8 is EOL and will not receive any more updates (except for potential security fixes)

Waester commented:
Understood, I'll try my hand on unstable in the mean time.
